### DISCUSSION TITLE
Consider production ready for timestamp oracle?

### DISCUSSION THREAD

**Discussion by danthegoodman1** [databendlabs/openraft#1161] [Q&A] [answered]
I'm interested in making a timestamp oracle in rust, and looking at using this crate instead of tikv's raft as this looks far easier to use and forget about.

I'm curious on whether this would be something advised _against_. I know the API my change before 1.0.0, but I'm curious if the maintainers would have concerns about testing, examples lacking extra instrumentation details (e.g. is there something non-obvious I have to add to detect failures and such, esp with storage), etc.

**Comment by drmingdrmer** [comment, marked as answer]
Openraft provides a very basic `RaftLogStorage` implementation test to ensure the implementation works as expected.
https://github.com/datafuselabs/openraft/blob/3ae6b4bfe2d13e633e437373f6c1b716b167bfa3/stores/memstore/src/test.rs#L21-L24

But internal the implementation it can not be tested. An real world production should be very careful with the storage.
The doc of the `RaftLogStorage` explained the requirement the application must achieve.

For example, it is crucial that the storage system does not re-order any IO operations: if `save_vote()` is called before `append()`, then `save_vote()` must be flushed to disk first before `append()`. Otherwise, the consensus cannot be guaranteed. 

An obvious mistake would be spawning a `task` to handle the IO operation in `save_vote()` and spawning another IO `task` in `append()`. In this case, there is no guarantee which task will be flushed first.

### Timestamp Project and OpenRaft

OpenRaft is suitable for your timestamp project. However, to achieve optimal performance, some form of batching should be implemented on top of Raft. Using a single Raft log entry to generate one timestamp can only achieve approximately x*10 kilo TPS.

**Reply by danthegoodman1** [reply]
Thanks for this, I ended up using Go https://github.com/danthegoodman1/EpicEpoch because I was struggling to figure out the typing with the getting started guide (probably me being far newer to Rust). The guide was lacking the example of the Responder, and I couldn't quite figure it out.

I would have really liked to have used this in Rust though!

**Reply by drmingdrmer** [reply]
You do not have to use or understand `Responder` when starting up a minimized example.
Clone an example such as https://github.com/datafuselabs/openraft/tree/main/examples/raft-kv-memstore
and modify upon it would be the quickest way
